src/mono/wasm/runtime/binding_support.js

```diff
@@ -1660,78 +1619,66 @@ var BindingSupportLib = {
-								this.mono_wasm_free_list.pop() : this.mono_wasm_ref_counter++;
-					js_obj.__mono_jshandle__ = handle;
+					var js_handle = this.mono_wasm_free_list.length ? this.mono_wasm_free_list.pop() : this.mono_wasm_ref_counter++;
+					js_obj.__mono_js_handle__ = js_handle;
 					// Obtain the JS -> C# type mapping.
 					var wasm_type = js_obj[Symbol.for("wasm type")];
 					js_obj.__owns_handle__ = true;
-					gc_handle = js_obj.__mono_gchandle__ = this.wasm_binding_obj_new(handle + 1, js_obj.__owns_handle__, typeof wasm_type === "undefined" ? -1 : wasm_type);
-					this.mono_wasm_object_registry[handle] = js_obj;
-					// as this instance was just created, it was already created with Inflight strong GCHandle, so we do not have to do it again
+					gc_handle = js_obj.__mono_gc_handle__ = this._bind_js_obj(js_handle, js_obj.__owns_handle__, typeof wasm_type === "undefined" ? -1 : wasm_type);
+					this.mono_wasm_object_registry[js_handle] = js_obj;
+					// as this instance was just created, it was already created with Inflight strong gc_handle, so we do not have to do it again
 					return { gc_handle, should_add_in_flight: false };
 				}
 			}
-			// this is pre-existing instance, we need to add Inflight strong GCHandle before passing it to managed
+			// this is pre-existing instance, we need to add Inflight strong gc_handle before passing it to managed
 			return { gc_handle, should_add_in_flight: true };
 		},
-		mono_wasm_require_handle: function(handle) {
-			if (handle > 0)
-				return this.mono_wasm_object_registry[handle - 1];
+		// TODO rename me
+		mono_wasm_require_handle: function(js_handle) {
+			if (js_handle > 0)
+				return this.mono_wasm_object_registry[js_handle];
 			return null;
 		},
-		mono_wasm_unregister_obj: function(js_id) {
-			var obj = this.mono_wasm_object_registry[js_id - 1];
+		mono_wasm_unregister_obj: function(js_handle, is_exception) {
+			var obj = BINDING.mono_wasm_object_registry[js_handle];
 			if (typeof obj  !== "undefined" && obj !== null) {
 				// if this is the global object then do not
 				// unregister it.
 				if (globalThis === obj)
 					return obj;
 
-				var gc_handle = obj.__mono_gchandle__;
-				if (typeof gc_handle  !== "undefined") {
-
-					obj.__mono_gchandle__ = undefined;
-					obj.__mono_jshandle__ = undefined;
-
-					// If we are unregistering a delegate then mark it as not being alive
-					//  so that attempts will not be made to invoke it even if a JS-side
-					//  reference to it remains (registered as an event handler, etc)
-					if (typeof obj.__mono_delegate_alive__ !== "undefined")
-						obj.__mono_delegate_alive__ = false;
-
-					this.mono_wasm_object_registry[js_id - 1] = undefined;
-					this.mono_wasm_free_list.push(js_id - 1);
+				if (typeof obj.__mono_gc_handle__  !== "undefined") {
+					obj.__mono_gc_handle__ = undefined;
+					obj.__mono_js_handle__ = undefined;
 				}
+
+				delete BINDING.mono_wasm_object_registry[js_handle];
```

Comment:
delete on js arrays is a horrible idea. never do it

Reply:
Ouch, I have not realized it's not `Map`, reverted